src/Common/ConnectionPool.cpp

```diff
@@ -0,0 +1,1075 @@
+            {
+                LOG_WARNING(log, "Too many active sessions in group {}, count {}", type, active_connections_in_group);
+                mute_warning_until = roundUp(active_connections_in_group, 100);
+            }
+        }
+
+        void atConnectionDestroy()
+        {
+            std::lock_guard lock(mutex);
+
+            --active_connections_in_group;
+
+            if (active_connections_in_group < limits.warning_limit && mute_warning_until > 0)
+            {
+                LOG_WARNING(log, "Sessions count is OK in the group {}, count {}", type, active_connections_in_group);
+                mute_warning_until = 0;
+            }
+        }
+
+        void addConnections(size_t num)
+        {
+            std::lock_guard lock(mutex);
+            active_connections_in_group += num;
+        }
+
+        void subConnections(size_t num)
+        {
+            std::lock_guard lock(mutex);
+            if (num > active_connections_in_group)
+                active_connections_in_group -= num;
+            else
+                active_connections_in_group = 0;
+        }
+
+        DB::ConnectionGroupType getType() const
+        {
+            return type;
+        }
+
+        const DB::ConnectionPoolMetrics & getMetrics() const
+        {
+            return metrics;
+        }
+
+    private:
+
+        const DB::ConnectionGroupType type;
+        const DB::ConnectionPoolMetrics metrics;
+
+        LoggerPtr log = ::getLogger("ConnectionGroup");
+
+        mutable std::mutex mutex;
+        DB::ConnectionPoolLimits limits TSA_GUARDED_BY(mutex) = DB::ConnectionPoolLimits();
+        size_t active_connections_in_group TSA_GUARDED_BY(mutex) = 0;
+        size_t mute_warning_until TSA_GUARDED_BY(mutex) = 0;
+    };
+
+
+    class IExtendedPool : public DB::IEndpointConnectionPool
+    {
+    public:
+        using Ptr = std::shared_ptr<IExtendedPool>;
+
+        virtual DB::ConnectionGroupType getGroupType() const = 0;
+        virtual void setGroup(ConnectionGroup::Ptr new_group) = 0;
+        virtual void dropResolvedHostsCache() = 0;
+        virtual size_t wipeExpired() = 0;
+    };
+
+
+    // EndpointConnectionPool manage connections to the endpoint
+    // Features:
+    // - it uses HostResolvePool for address selecting. See Common/HostResolvePool.h for more info.
+    // - it minimizes number of `Session::connect()`/`Session::reconnect()` calls
+    //   - stores only connected and ready to use sessions
+    //   - connection could be reused even when limits are reached
+    // - soft limit, warn limit, hard limit
+    // - `Session::reconnect()` uses the pool as well
+    // - comprehensive sensors
+    // - session is reused according its inner state, automatically
+
+
+    // connection live stages
+    // CREATED -> STORED
+    // CREATED -> RESET
+    // STORED -> EXPIRED
+    // STORED -> REUSED
+    // REUSED -> RESET
+    // REUSED -> STORED
```

Comment:
Nitpick: this confused me more than it helped.
 * It looks like a state transition diagram, but the words sound like actions rather than states. I would've expected something like `IDLE <-> CONNECTED <-> DISCONNECTED` or whatever, maybe with things like "connect"/"reuse"/"error"/"expire" written on the arrows.
 * Is `CREATED -> STORED` possible? Seems that we only create sessions on demand and immediately use them.
 * What's the difference between CREATED and REUSED? I.e. does the connection care whether it's being used for the first or second time?
 * Maybe the trickiest part of PooledConnection is jumpToOtherConnection(), but this diagram doesn't mention it. Does jumpToOtherConnection() do a REUSED to this connection and do a RESET to the other connection? And STORED actually consists of creating a new PooledConnection, doing a jumpToOtherConnection() on it, then discarding the old PooledConnection, so from perspective of one PooledConnection it's more of a reset, and from perspective of the other it's more of a create, or something, thinking about state transition diagrams doesn't seem to help here.

I mean, it probably makes sense as is, just by the time I understood what it may mean I already knew how everything works, and it wasn't useful.

I guess state transition diagram for PooledConnection would look something more like this:
```
                      +---------------------+  +--expire--> EMPTY
                      v                     ^  ^
EMPTY --prepare--> IN_USE --+--complete--> STORED  (jumped into a new PooledConnection instance, old instance becomes EMPTY and destroyed)
                            +--complete--> EMPTY  (decided to not reuse)
                            +--error--> EMPTY
                            +--disconnected--> IN_USE  (jumped from another PooledConnection, which goes from STORED to EMPTY and destroyed)
```
which is still is not very accurate and probably worse than nothing

Reply:
I just removed that diagram. Collecting stats code helps here more as an logical hint.